File: backend/src/main/java/com/nhtl/models/Commande.java
package com.nhtl.models;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "commandes")
public class Commande {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String userId;

    private String nom;
    private String prenom;
    private String numeroTelephone;
    private String email;

    private String paysLivraison;
    private String villeLivraison;
    @Column(columnDefinition = "TEXT")
    private String adresseLivraison;

    private String plateforme;

    @Column(columnDefinition = "TEXT")
    private String lienProduit;

    @Column(name = "liens_produits", columnDefinition = "TEXT")
    private String liensProduits;

    @Column(name = "photos_produits", columnDefinition = "TEXT")
    private String photosProduits;

    @Column(name = "articles_json", columnDefinition = "TEXT")
    private String articlesJson;

    @Column(columnDefinition = "TEXT")
    private String descriptionCommande;

    private Integer quantite;
    private BigDecimal prixUnitaire;
    private BigDecimal prixTotal;
    private String devise;
    @Column(columnDefinition = "TEXT")
    private String notesSpeciales;

    // Statut ADMINISTRATIF (gestion du dossier)
    // Valeurs : EN_ATTENTE, EN_COURS, LIVRE, ANNULE
    private String statut = "EN_ATTENTE";

    // ✅ Statut LOGISTIQUE (suivi de la livraison physique)
    @Enumerated(EnumType.STRING)
    @Column(name = "statut_suivi_commande", nullable = false)
    private CommandeStatus statutSuivi = CommandeStatus.EN_ATTENTE;

    @Column(nullable = false)
    private Boolean archived = false;

    // GP assignment
    @Column(name = "gp_id")
    private Long gpId;
    @Column(name = "gp_prenom")
    private String gpPrenom;
    @Column(name = "gp_nom")
    private String gpNom;
    @Column(name = "gp_phone_number")
    private String gpPhoneNumber;

    private LocalDateTime dateCreation;
    private LocalDateTime dateModification;

    @PrePersist
    protected void onCreate() {
        dateCreation     = LocalDateTime.now();
        dateModification = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        dateModification = LocalDateTime.now();
    }

    // ── Helpers ────────────────────────────────────────────────────────────

    public String getReference() {
        return "#" + id + " — " + (plateforme != null ? plateforme : "?")
                + " → " + (paysLivraison != null ? paysLivraison : "?");
    }

    public String getClientFullName() {
        String p = prenom != null ? prenom : "";
        String n = nom    != null ? nom    : "";
        return (p + " " + n).trim();
    }

    // ── Getters & Setters ──────────────────────────────────────────────────

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getUserId() { return userId; }
    public void setUserId(String userId) { this.userId = userId; }

    public String getNom() { return nom; }
    public void setNom(String nom) { this.nom = nom; }

    public String getPrenom() { return prenom; }
    public void setPrenom(String prenom) { this.prenom = prenom; }

    public String getNumeroTelephone() { return numeroTelephone; }
    public void setNumeroTelephone(String n) { this.numeroTelephone = n; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getPaysLivraison() { return paysLivraison; }
    public void setPaysLivraison(String p) { this.paysLivraison = p; }

    public String getVilleLivraison() { return villeLivraison; }
    public void setVilleLivraison(String v) { this.villeLivraison = v; }

    public String getAdresseLivraison() { return adresseLivraison; }
    public void setAdresseLivraison(String a) { this.adresseLivraison = a; }

    public String getPlateforme() { return plateforme; }
    public void setPlateforme(String p) { this.plateforme = p; }

    public String getLienProduit() { return lienProduit; }
    public void setLienProduit(String l) { this.lienProduit = l; }

    public String getLiensProduits() { return liensProduits; }
    public void setLiensProduits(String l) { this.liensProduits = l; }

    public String getPhotosProduits() { return photosProduits; }
    public void setPhotosProduits(String p) { this.photosProduits = p; }

    public String getArticlesJson() { return articlesJson; }
    public void setArticlesJson(String a) { this.articlesJson = a; }

    public String getDescriptionCommande() { return descriptionCommande; }
    public void setDescriptionCommande(String d) { this.descriptionCommande = d; }

    public Integer getQuantite() { return quantite; }
    public void setQuantite(Integer q) { this.quantite = q; }

    public BigDecimal getPrixUnitaire() { return prixUnitaire; }
    public void setPrixUnitaire(BigDecimal p) { this.prixUnitaire = p; }

    public BigDecimal getPrixTotal() { return prixTotal; }
    public void setPrixTotal(BigDecimal p) { this.prixTotal = p; }

    public String getDevise() { return devise; }
    public void setDevise(String d) { this.devise = d; }

    public String getNotesSpeciales() { return notesSpeciales; }
    public void setNotesSpeciales(String n) { this.notesSpeciales = n; }

    public String getStatut() { return statut; }
    public void setStatut(String s) { this.statut = s; }

    public CommandeStatus getStatutSuivi() { return statutSuivi; }
    public void setStatutSuivi(CommandeStatus s) { this.statutSuivi = s; }

    public Boolean getArchived() { return archived; }
    public void setArchived(Boolean a) { this.archived = a; }

    public Long getGpId() { return gpId; }
    public void setGpId(Long g) { this.gpId = g; }

    public String getGpPrenom() { return gpPrenom; }
    public void setGpPrenom(String g) { this.gpPrenom = g; }

    public String getGpNom() { return gpNom; }
    public void setGpNom(String g) { this.gpNom = g; }

    public String getGpPhoneNumber() { return gpPhoneNumber; }
    public void setGpPhoneNumber(String g) { this.gpPhoneNumber = g; }

    public LocalDateTime getDateCreation() { return dateCreation; }
    public void setDateCreation(LocalDateTime d) { this.dateCreation = d; }

    public LocalDateTime getDateModification() { return dateModification; }
    public void setDateModification(LocalDateTime d) { this.dateModification = d; }

    // ── Suivi postal ──────────────────────────────────────────────────────────
    @Column(name = "photo_colis_url")
    private String photoColisUrl;

    @Column(name = "photo_bordereau_url")
    private String photoBordereauUrl;

    @Column(name = "numero_bordereau", length = 100)
    private String numeroBordereau;

    @Column(name = "depose_poste_at")
    private LocalDateTime deposePosteAt;

    public String getPhotoColisUrl() { return photoColisUrl; }
    public void setPhotoColisUrl(String u) { this.photoColisUrl = u; }

    public String getPhotoBordereauUrl() { return photoBordereauUrl; }
    public void setPhotoBordereauUrl(String u) { this.photoBordereauUrl = u; }

    public String getNumeroBordereau() { return numeroBordereau; }
    public void setNumeroBordereau(String n) { this.numeroBordereau = n; }

    public LocalDateTime getDeposePosteAt() { return deposePosteAt; }
    public void setDeposePosteAt(LocalDateTime d) { this.deposePosteAt = d; }

    public boolean isDeposePoste() { return deposePosteAt != null; }
}

File: frontend/lib/models/commande.dart
// lib/models/commande.dart

class Commande {
  final int? id;
  final String nom;
  final String prenom;
  final String numeroTelephone;
  final String? email;
  final String paysLivraison;
  final String villeLivraison;
  final String adresseLivraison;
  final String plateforme;
  final String lienProduit;
  final String descriptionCommande;
  final int quantite;
  final double prixUnitaire;
  final double prixTotal;
  final String devise;
  final String? notesSpeciales;

  // Statut ADMINISTRATIF (gestion du dossier)
  // Valeurs : EN_ATTENTE, EN_COURS, LIVRE, ANNULE
  final String statut;

  // ✅ Statut LOGISTIQUE (suivi physique de la livraison)
  // Valeurs : EN_ATTENTE, COMMANDE_CONFIRMEE, EN_TRANSIT, EN_DOUANE,
  //           ARRIVE, PRET_LIVRAISON, LIVRE
  final String statutSuivi;

  final bool archived;
  final String? userId;
  final DateTime? dateCreation;
  final DateTime? dateModification;

  // GP assignment
  final int? gpId;
  final String? gpPrenom;
  final String? gpNom;
  final String? gpPhoneNumber;

  Commande({
    this.id,
    required this.nom,
    required this.prenom,
    required this.numeroTelephone,
    this.email,
    required this.paysLivraison,
    required this.villeLivraison,
    required this.adresseLivraison,
    required this.plateforme,
    required this.lienProduit,
    required this.descriptionCommande,
    required this.quantite,
    required this.prixUnitaire,
    required this.prixTotal,
    this.devise = 'EUR',
    this.notesSpeciales,
    this.statut = 'EN_ATTENTE',
    this.statutSuivi = 'EN_ATTENTE',
    this.archived = false,
    this.userId,
    this.dateCreation,
    this.dateModification,
    this.gpId,
    this.gpPrenom,
    this.gpNom,
    this.gpPhoneNumber,
  });

  Commande copyWith({
    int? id,
    String? nom,
    String? prenom,
    String? numeroTelephone,
    String? email,
    String? paysLivraison,
    String? villeLivraison,
    String? adresseLivraison,
    String? plateforme,
    String? lienProduit,
    String? descriptionCommande,
    int? quantite,
    double? prixUnitaire,
    double? prixTotal,
    String? devise,
    String? notesSpeciales,
    String? statut,
    String? statutSuivi,
    bool? archived,
    String? userId,
    DateTime? dateCreation,
    DateTime? dateModification,
    int? gpId,
    String? gpPrenom,
    String? gpNom,
    String? gpPhoneNumber,
  }) {
    return Commande(
      id: id ?? this.id,
      nom: nom ?? this.nom,
      prenom: prenom ?? this.prenom,
      numeroTelephone: numeroTelephone ?? this.numeroTelephone,
      email: email ?? this.email,
      paysLivraison: paysLivraison ?? this.paysLivraison,
      villeLivraison: villeLivraison ?? this.villeLivraison,
      adresseLivraison: adresseLivraison ?? this.adresseLivraison,
      plateforme: plateforme ?? this.plateforme,
      lienProduit: lienProduit ?? this.lienProduit,
      descriptionCommande: descriptionCommande ?? this.descriptionCommande,
      quantite: quantite ?? this.quantite,
      prixUnitaire: prixUnitaire ?? this.prixUnitaire,
      prixTotal: prixTotal ?? this.prixTotal,
      devise: devise ?? this.devise,
      notesSpeciales: notesSpeciales ?? this.notesSpeciales,
      statut: statut ?? this.statut,
      statutSuivi: statutSuivi ?? this.statutSuivi,
      archived: archived ?? this.archived,
      userId: userId ?? this.userId,
      dateCreation: dateCreation ?? this.dateCreation,
      dateModification: dateModification ?? this.dateModification,
      gpId: gpId ?? this.gpId,
      gpPrenom: gpPrenom ?? this.gpPrenom,
      gpNom: gpNom ?? this.gpNom,
      gpPhoneNumber: gpPhoneNumber ?? this.gpPhoneNumber,
    );
  }

  factory Commande.fromJson(Map<String, dynamic> json) {
    return Commande(
      id: (json['id'] as num?)?.toInt(),
      nom: json['nom'] as String? ?? '',
      prenom: json['prenom'] as String? ?? '',
      numeroTelephone: json['numeroTelephone'] as String? ?? '',
      email: json['email'] as String?,
      paysLivraison: json['paysLivraison'] as String? ?? '',
      villeLivraison: json['villeLivraison'] as String? ?? '',
      adresseLivraison: json['adresseLivraison'] as String? ?? '',
      plateforme: json['plateforme'] as String? ?? '',
      lienProduit: json['lienProduit'] as String? ?? '',
      descriptionCommande: json['descriptionCommande'] as String? ?? '',
      quantite: (json['quantite'] as num?)?.toInt() ?? 1,
      prixUnitaire: (json['prixUnitaire'] is int)
          ? (json['prixUnitaire'] as int).toDouble()
          : (json['prixUnitaire'] as double? ?? 0.0),
      prixTotal: (json['prixTotal'] is int)
          ? (json['prixTotal'] as int).toDouble()
          : (json['prixTotal'] as double? ?? 0.0),
      devise: json['devise'] as String? ?? 'EUR',
      notesSpeciales: json['notesSpeciales'] as String?,
      statut: json['statut'] as String? ?? 'EN_ATTENTE',
      // ✅ statutSuivi logistique — défaut EN_ATTENTE si absent
      statutSuivi: json['statutSuivi'] as String? ?? 'EN_ATTENTE',
      archived: json['archived'] is int
          ? (json['archived'] == 1)
          : (json['archived'] as bool? ?? false),
      userId: json['userId'] as String?,
      dateCreation: json['dateCreation'] != null
          ? DateTime.tryParse(json['dateCreation'].toString())
          : null,
      dateModification: json['dateModification'] != null
          ? DateTime.tryParse(json['dateModification'].toString())
          : null,
      gpId: (json['gpId'] as num?)?.toInt(),
      gpPrenom: json['gpPrenom'] as String?,
      gpNom: json['gpNom'] as String?,
      gpPhoneNumber: json['gpPhoneNumber'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nom': nom,
      'prenom': prenom,
      'numeroTelephone': numeroTelephone,
      'email': email,
      'paysLivraison': paysLivraison,
      'villeLivraison': villeLivraison,
      'adresseLivraison': adresseLivraison,
      'plateforme': plateforme,
      'lienProduit': lienProduit,
      'descriptionCommande': descriptionCommande,
      'quantite': quantite,
      'prixUnitaire': prixUnitaire,
      'prixTotal': prixTotal,
      'devise': devise,
      'notesSpeciales': notesSpeciales,
      'statut': statut,
      // statutSuivi géré par PATCH /api/admin/commandes/{id}/status
      'gpId': gpId,
      'gpPrenom': gpPrenom,
      'gpNom': gpNom,
      'gpPhoneNumber': gpPhoneNumber,
    };
  }
}
